### PR TITLE
Create @firebase/app-compat and firebase/compat

### PR DESCRIPTION
**Pull request by Feiyang1**

### CODE REVIEW

File: packages-exp/app-compat/src/firebaseNamespaceCore.ts [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    return apps[name];
+  }
+
+  // @ts-ignore
+  app['App'] = firebaseAppImpl;
+
+  /**
+   * Create a new App instance (name must be unique).
+   */
+  function initializeAppCompat(
```
Comment (hsubox76) [inline packages-exp/app-compat/src/firebaseNamespaceCore.ts:121]: Doesn't accept a string as a second arg anymore?

File: packages-exp/app-compat/src/firebaseNamespaceCore.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  // A namespace is a plain JavaScript Object.
+  const namespace: FirebaseNamespace = {
+    // Hack to prevent Babel from modifying the object returned
+    // as the firebase namespace.
+    // @ts-ignore
+    __esModule: true,
+    initializeApp: initializeAppCompat,
+    // @ts-ignore
+    app,
+    registerVersion,
```
Comment (hsubox76) [inline packages-exp/app-compat/src/firebaseNamespaceCore.ts:67]: No more `onLog()` and `setLogLevel()`?

File: packages-exp/app-compat/src/lite/firebaseAppLite.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+  get name(): string {
+    return this.app.name;
+  }
+
+  get options(): FirebaseOptions {
+    return this.app.options;
+  }
+
+  delete(): Promise<void> {
```
Comment (hsubox76) [inline packages-exp/app-compat/src/lite/firebaseAppLite.ts:61]: Why is this one 2 lines while the one in firebaseApp has a 2 step Promise with a checkDestroyed check?

File: packages-exp/app-compat/test/firebaseCompat.test.ts [hunk 4]
```diff
@@ -0,0 +1,365 @@
+/**
```
Comment (hsubox76) [inline packages-exp/app-compat/test/firebaseCompat.test.ts:1]: name: firebaseAppCompat.test.ts ?

File: packages-exp/firebase-exp/compat/index.cdn.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+(replace <PACKAGE> with the name of a component - i.e. auth, database, etc):
+
+https://www.gstatic.com/firebasejs/5.0.0/firebase-<PACKAGE>.js
+`);
+
+import '@firebase/polyfill';
+import firebase from './app';
+import { name, version } from '../package.json';
+
+// import './auth';
```
Comment (hsubox76) [inline packages-exp/firebase-exp/compat/index.cdn.ts:33]: What's going to happen with these? No more all-in-one bundles? If not, do we need the warning anymore? Or is this just temporary during development?

File: packages-exp/app-compat/src/firebaseNamespaceCore.ts (hunk 1 above)
Comment (Feiyang1) [reply packages-exp/app-compat/src/firebaseNamespaceCore.ts:121]: It does. `FirebaseNamespace` defines the signature for `firebase.initializeApp` which I didn't change.

File: packages-exp/app-compat/src/lite/firebaseAppLite.ts (hunk 3 above)
Comment (Feiyang1) [reply packages-exp/app-compat/src/lite/firebaseAppLite.ts:61]: Just to make the light version smaller by sacrificing a little safety. I'm actually not sure if we still need a lite build for compat layer. I will keep it for now, but might delete it in the future.

File: packages-exp/app-compat/test/firebaseCompat.test.ts (hunk 4 above)
Comment (Feiyang1) [reply packages-exp/app-compat/test/firebaseCompat.test.ts:1]: Renamed.

File: packages-exp/firebase-exp/compat/index.cdn.ts (hunk 5 above)
Comment (Feiyang1) [reply packages-exp/firebase-exp/compat/index.cdn.ts:33]: We will add them as their compat packages become available. We will have a compat all-in-one bundle.

File: packages/firestore/exp/src/api/components.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ import {
   OfflineComponentProvider,
   OnlineComponentProvider
 } from '../../../src/core/component_provider';
-import {handleUserChange, LocalStore} from '../../../src/local/local_store';
+import { handleUserChange, LocalStore } from '../../../src/local/local_store';
```
Comment (hsubox76) [inline packages/firestore/exp/src/api/components.ts:26]: Looks like this PR includes some prettier fixes not caught during firestore or rules-unit-testing commits somehow, do you want to include them here or in a separate PR?

**Pull request merged**